pkg/api/rest/request.go

```diff
@@ -0,0 +1,397 @@
+	r.timeout = d
+	return r
+}
+
+// BackoffBaseDelay specifies the base backoff sleep duration.
+func (r *Request) BackoffBaseDelay(delay time.Duration) *Request {
+	if r.err != nil {
+		return r
+	}
+	r.backoffBaseDelay = delay
+	return r
+}
+
+// BackoffMaxDelay specifies the maximum backoff sleep duration.
+func (r *Request) BackoffMaxDelay(delay time.Duration) *Request {
+	if r.err != nil {
+		return r
+	}
+	r.backoffMaxDelay = delay
+	return r
+}
+
+// MaxRetries specifies the maximum times a request will retry.
+func (r *Request) MaxRetries(maxRetries int64) *Request {
+	if r.err != nil {
+		return r
+	}
+	if maxRetries > 0 {
+		r.maxRetries = maxRetries
+	} else {
+		r.maxRetries = defaultMaxRetries
+	}
+	return r
+}
+
+// Body makes http request use obj as its body.
+func (r *Request) Body(obj interface{}) *Request {
+	if r.err != nil {
+		return r
+	}
+
+	// only supports two types now:
+	//   1. io.Reader
+	//   2. type which can be json marshalled
+	if rd, ok := obj.(io.Reader); ok {
+		r.body = rd
+	} else {
+		b, err := json.Marshal(obj)
+		if err != nil {
+			r.err = err
+			return r
+		}
+		r.body = bytes.NewReader(b)
+		r.SetHeader("Content-Type", "application/json")
+	}
+	return r
+}
+
+// URL returns the current working URL.
+func (r *Request) URL() *url.URL {
+	p := r.pathPrefix
+
+	finalURL := &url.URL{}
+	if r.c.base != nil {
+		*finalURL = *r.c.base
+	}
+	finalURL.Path = p
+
+	query := url.Values{}
+	for key, values := range r.params {
+		for _, value := range values {
+			query.Add(key, value)
+		}
+	}
+
+	finalURL.RawQuery = query.Encode()
+	return finalURL
+}
+
+func (r *Request) newHttpRequest(ctx context.Context) (*http.Request, error) {
+	url := r.URL().String()
+	req, err := http.NewRequest(r.verb, url, r.body)
+	if err != nil {
+		return nil, err
+	}
+	req = req.WithContext(ctx)
+	req.Header = r.headers
+	return req, nil
+}
+
+// Do formats and executes the request.
+func (r *Request) Do(ctx context.Context) (res *Result) {
+	if r.err != nil {
+		log.Info("error in request", zap.Error(r.err))
+		return &Result{err: r.err}
+	}
+
+	client := r.c.Client
+	if client == nil {
+		client = &httputil.Client{}
+	}
+
+	if r.timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, r.timeout)
+		defer cancel()
+	}
+
+	baseDelay := r.backoffBaseDelay.Milliseconds()
+	if baseDelay == 0 {
+		baseDelay = defaultBackoffBaseDelayInMs
+	}
+	maxDelay := r.backoffMaxDelay.Milliseconds()
+	if maxDelay == 0 {
+		maxDelay = defaultBackoffMaxDelayInMs
+	}
+	maxRetries := r.maxRetries
+	if maxRetries == 0 {
+		maxRetries = defaultMaxRetries
+	}
+
+	err := retry.Do(ctx, func() error {
+		req, err := r.newHttpRequest(ctx)
+		if err != nil {
+			return err
+		}
+		// rewind the request body when r.body is not nil
+		if seeker, ok := r.body.(io.Seeker); ok && r.body != nil {
+			if _, err := seeker.Seek(0, 0); err != nil {
+				return terrors.Annotatef(err, "failed to seek to the beginning of request body:%v", r)
```

Comment:
Should we use `pkg/errors`?

Reply:
> Should we use `pkg/errors`?

Do you mean this error should be added into pkg/errors ?

Reply:
Yes